.github/workflows/macos.yml

```diff
@@ -0,0 +1,269 @@
+name: macOS-Builds
+
+on:
+  workflow_call:
+    inputs:
+      checkout-ref:
+        required: true
+        description: "The ref we want to compile"
+        type: string
+
+permissions:
+  contents: read
+
+jobs:
+  bazel:
+    name: bazel + ${{ matrix.os }} + ${{ matrix.shard }}
+    runs-on: ${{ matrix.os }}-xl
+    permissions:
+      contents: 'read'
+      id-token: 'write'
+    strategy:
+      # Continue other builds even if one fails
+      fail-fast: false
+      matrix:
+        os: [ macos-12 ]
+        shard: [ 'Core', 'Compute', 'Larger Libraries', 'Other' ]
+        include:
+        - shard: Core
+          targets:
+          - //google/cloud:all
+          # - //generator/...  # Does not build on macOS
+          # - //docfx/...      # Does not build on macOS
+          - +//google/cloud/bigtable/...
+          - +//google/cloud/pubsub/...
+          - +//google/cloud/spanner/...
+          - +//google/cloud/storage/...
+          # Include the top-level examples because they require the libs anyway.
+          - +//examples/...
+        - shard: Compute
+          targets:
+          - //google/cloud/compute/...
+        - shard: Larger Libraries
+          targets:
+          # Run this:
+          #
+          # for file in google/cloud/*/v*/*_client.h; do
+          #     echo $(dirname $(dirname $file));
+          # done | sort | uniq -c | sort -n  | awk '{ s += $1; print s, $0}'
```

Comment:
nit: This will miss services like `bigquery` (nested directories) and `dialogflow_*` (not yet consolidated). I think that will throw off your balancing a bit.

e.g. if I run:
```sh
grep -r "class.*Client" --include "*_client.h" $(git ls-files google/cloud -- .h) -l | egrep -v "(bigtable/|internal/|pubsub/|spanner/|storage/)" | cut -f -3 -d/ | sort | uniq -c | sort -n | awk '{ s += $1; print s, $0}'
```

We can get a smaller (and more balanced?) subset:
```
125       3 google/cloud/servicecontrol
128       3 google/cloud/support
131       3 google/cloud/video
135       4 google/cloud/datacatalog
139       4 google/cloud/iam
143       4 google/cloud/kms
148       5 google/cloud/beyondcorp
153       5 google/cloud/contentwarehouse
158       5 google/cloud/dataproc
163       5 google/cloud/talent
169       6 google/cloud/retail
176       7 google/cloud/bigquery
184       8 google/cloud/appengine
193       9 google/cloud/monitoring
202       9 google/cloud/sql
218      16 google/cloud/dialogflow_cx
235      17 google/cloud/aiplatform
252      17 google/cloud/dialogflow_es
```